src/Properties/JsonResourceExtension.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace NunoMaduro\Larastan\Properties;
+
+use Illuminate\Http\Resources\Json\JsonResource;
+use PHPStan\Analyser\OutOfClassScope;
+use PHPStan\Reflection\ClassReflection;
+use PHPStan\Reflection\Dummy\DummyMethodReflection;
+use PHPStan\Reflection\Dummy\DummyPropertyReflection;
+use PHPStan\Reflection\MethodsClassReflectionExtension;
+use PHPStan\Reflection\PropertiesClassReflectionExtension;
+use PHPStan\Type\ObjectType;
+
+class JsonResourceExtension implements PropertiesClassReflectionExtension, MethodsClassReflectionExtension
+{
+    public function hasMethod(ClassReflection $classReflection, string $methodName): bool
+    {
+        if (! $classReflection->isSubclassOf(JsonResource::class)) {
+            return false;
+        }
+
+        $modelClassReflection = $this->getModelClassReflection($classReflection);
+
+        if (! $modelClassReflection) {
+            return false;
+        }
+
+        return $modelClassReflection->hasMethod($methodName);
+    }
+
+    public function getMethod(ClassReflection $classReflection, string $methodName): \PHPStan\Reflection\MethodReflection
+    {
+        $modelClassReflection = $this->getModelClassReflection($classReflection);
+
+        if (! $modelClassReflection) {
+            return new DummyMethodReflection($methodName);
+        }
+
+        return $modelClassReflection->getMethod($methodName, new OutOfClassScope());
+    }
+
+    public function hasProperty(ClassReflection $classReflection, string $propertyName): bool
+    {
+        if (! $classReflection->isSubclassOf(JsonResource::class)) {
+            return false;
+        }
+
+        $modelClassReflection = $this->getModelClassReflection($classReflection);
+
+        if (! $modelClassReflection) {
+            return false;
+        }
+
+        return $modelClassReflection->hasProperty($propertyName);
+    }
+
+    public function getProperty(ClassReflection $classReflection, string $propertyName): \PHPStan\Reflection\PropertyReflection
+    {
+        $modelClassReflection = $this->getModelClassReflection($classReflection);
+
+        if (! $modelClassReflection) {
+            return new DummyPropertyReflection();
+        }
+
+        return $modelClassReflection->getProperty($propertyName, new OutOfClassScope());
+    }
+
+    private function getModelClassReflection(ClassReflection $classReflection): ?ClassReflection
+    {
+        $templateTypeMap = $classReflection->getParents()[0]->getActiveTemplateTypeMap();
```

Comment:
Yes, that is what I originally tried. The test fails and there are no types returned in the template type map. Not sure how to proceed

Reply:
Aah. The docblock should be 

```php
/**
 * @template TModelClass
 * @extends JsonResource<User>
 */
```

I think. For the `UserResource` Can you try like that?

Reply:
I've been playing around, and can get `$classReflection->getActiveTemplateTypeMap();` to work if we change the docblock for end-users to implement as 

```
/**
 * @template TModelClass of User
 */
```

instead of 

```
/**
 * @extends JsonResource<User>
 */
```

Which do we prefer, or shall I try something else?

Reply:
```
/**
 * @template TModelClass of User
 */
```

will make the child class also generic. I think we don't want this. So the current way is fine.